dask_ml/preprocessing/_encoders.py

```diff
@@ -0,0 +1,278 @@
+
+        if is_array:
+            for i in range(n_features):
+                Xi = X[:, i]
+                if self.categories == "auto":
+                    cats = _encode(Xi)
+                else:
+                    cats = np.array(self.categories[i], dtype=X.dtype)
+                    if self.handle_unknown == "error":
+                        # TODO: check unknown
+                        # diff = _encode_check_unknown(Xi, cats)
+                        # if diff:
+                        #     msg = (
+                        #         "Found unknown categories {0} in column {1}"
+                        #         " during fit".format(diff, i)
+                        #     )
+                        #     raise ValueError(msg)
+                        pass
+                self.categories_.append(cats)
+                self.dtypes_.append(None)
+        else:
+            if not (X.dtypes == "category").all():
+                raise ValueError("All columns must be Categorical dtype.")
+            if self.categories == "auto":
+                for col in X.columns:
+                    Xi = X[col]
+                    cats = _encode(Xi, uniques=Xi.cat.categories)
+                    self.categories_.append(cats)
+                    self.dtypes_.append(Xi.dtype)
+            else:
+                raise ValueError(
+                    "Cannot specify 'categories' with DataFrame input. "
+                    "Use a categorical dtype instead."
+                )
+
+        self.categories_ = dask.compute(self.categories_)[0]
+
+    def _transform_new(self, X):
+        # TODO: check_array
+        is_array = isinstance(X, da.Array)
+
+        if is_array:
+            _, n_features = X.shape
+        else:
+            n_features = len(X.columns)
+
+        if is_array:
+            # For arrays, the basic plan is:
+            # User-array -> for each column:
+            #     encoded dask array ->
+            #     List[Delayed[sparse]] ->
+            #     Array[sparse] -> (optionally Array[Dense]).
+            Xs = [
+                _encode(X[:, i], self.categories_[i], encode=True)[1]
+                for i in range(n_features)
+            ]
+
+            objs = [
+                [dask.delayed(construct)(block, categories) for block in x.to_delayed()]
+                for x, categories in zip(Xs, self.categories_)
+            ]
+            arrs = []
+
+            for i, (objs, x, categories) in enumerate(zip(objs, Xs, self.categories_)):
+                inner_ars = [
+                    # TODO: dtype
+                    da.from_delayed(obj, (n_rows, len(categories)), dtype=self.dtype)
+                    for j, (obj, n_rows) in enumerate(zip(objs, x.chunks[0]))
+                ]
+                arrs.append(da.concatenate(inner_ars))
+                # TODO: this check fails. See why.
+                # da.utils._check_dsk(arrs[-1].dask)
+
+            X = da.concatenate(arrs, axis=1)
```

Comment:
What stops us from using `X.map_blocks(construct, categories=categories) here?

Reply:
And removing all of the to / from delayed, instead doing it all in construct (but not using delayed)?

I think that should be doable...

Reply:
Yes.  My hope is that it would be a bit simpler.  There is also some cost to switching between collection types.  I think that every such boundary triggers an optimization step, which also stops us from fusing successive operations.